samza-azure/src/main/java/org/apache/samza/AzureJobCoordinator.java

```diff
@@ -0,0 +1,483 @@
+    return () -> {
+      LOG.info("Job model version upgrade detected.");
+      versionUpgradeDetected.getAndSet(true);
+      onNewJobModelAvailable(leaderBlob.getJobModelVersion());
+    };
+  }
+
+  /**
+   * Creates a listener for LeaderLivenessCheckScheduler class.
+   * Invoked when an existing leader dies. Enables the JC to participate in leader election again.
+   * @return an instance of SchedulerStateChangeListener.
+   */
+  private SchedulerStateChangeListener createLeaderLivenessListener() {
+    return () -> {
+      LOG.info("Existing leader died.");
+      azureLeaderElector.tryBecomeLeader();
+    };
+  }
+
+  /**
+   * For each input stream specified in config, exactly determine its
+   * partitions, returning a set of SystemStreamPartitions containing them all.
+   */
+  private Set<SystemStreamPartition> getInputStreamPartitions() {
+    TaskConfig taskConfig = new TaskConfig(config);
+    scala.collection.immutable.Set<SystemStream> inputSystemStreams = taskConfig.getInputStreams();
+
+    // Get the set of partitions for each SystemStream from the stream metadata
+    Set<SystemStreamPartition>
+        sspSet = JavaConverters.mapAsJavaMapConverter(streamMetadataCache.getStreamMetadata(inputSystemStreams, true)).asJava()
+        .entrySet()
+        .stream()
+        .flatMap(this::mapSSMToSSP)
+        .collect(Collectors.toSet());
+
+    return sspSet;
+  }
+
+  private Stream<SystemStreamPartition> mapSSMToSSP(Map.Entry<SystemStream, SystemStreamMetadata> ssMs) {
+    return ssMs.getValue()
+        .getSystemStreamPartitionMetadata()
+        .keySet()
+        .stream()
+        .map(partition -> new SystemStreamPartition(ssMs.getKey(), partition));
+  }
+
+  /**
+   * Gets a SystemStreamPartitionGrouper object from the configuration.
+   */
+  private SystemStreamPartitionGrouper getSystemStreamPartitionGrouper() {
+    JobConfig jobConfig = new JobConfig(config);
+    String factoryString = jobConfig.getSystemStreamPartitionGrouperFactory();
+    SystemStreamPartitionGrouper grouper = Util.<SystemStreamPartitionGrouperFactory>getObj(factoryString).getSystemStreamPartitionGrouper(jobConfig);
+    return grouper;
+  }
+
+  private int getMaxNumTasks() {
+    // Do grouping to fetch TaskName to SSP mapping
+    Set<SystemStreamPartition> allSystemStreamPartitions = getInputStreamPartitions();
+    SystemStreamPartitionGrouper grouper = getSystemStreamPartitionGrouper();
+    Map<TaskName, Set<SystemStreamPartition>> groups = grouper.group(allSystemStreamPartitions);
+    LOG.info("SystemStreamPartitionGrouper " + grouper.toString() + " has grouped the SystemStreamPartitions into " + Integer.toString(groups.size()) +
+        " tasks with the following taskNames: {}", groups.keySet());
+    return groups.size();
+  }
+
+  /**
+   * Called only by the leader, either when the processor becomes the leader, or when the list of live processors changes.
```

Comment:
Same question here, what if the leader lost the lease to leaderBlob while doOnProcessorChange() is running? Any leader-specific actions should be stopped until the leader resumes the lease, or completely aborted if another process assume the leader position.